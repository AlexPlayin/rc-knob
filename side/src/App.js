import React from 'react'

import { H1, H2, Grid, Description } from './styled'
import PointerExample from './examples/Pointer'
import ScaleExample from './examples/Scale'
import ArcExample from './examples/Arc'
import ArcPointerExample from './examples/Arc_Pointer'
import ScalePointerExample from './examples/Scale_Pointer'
import ValueExample from './examples/Value'
import ArcPointerValueExample from './examples/Arc_Pointer_Value'
import ScaleScale from './examples/Scale_Scale'
import PointerCustom from './examples/Pointer_custom'
import ScaleCustom from './examples/Scale_custom'
import ScaleStyled from './examples/Scale_styled'
import ScaleStyled2 from './examples/Scale_styled_2'
import ArcPointerStyled from './examples/Arc_Pointer_styled'
import EventMouse from "./examples/Event_mouse";
import EventReadOnly from "./examples/Event_readOnly";
import EventNoWheel from "./examples/Event_nowheel";


import Filter from './filter'

export default () => {
    return (
        <div>
            <Filter />
            <H1>React Knob</H1>
            <H2>
                <span>Base Components</span>
            </H2>
            <Description>
                The base of all knobs is the <pre>Knob</pre> component which handles
                all of the user interaction and calculation of the knob value. The
                user can interact via drag'n drop, mousewheel and keyboard. The
                knobs are accessible by the <pre>tab</pre> key.
                <br /> The display part is splitted into the <pre>
                    Pointer
                </pre>, <pre>Scale</pre>, <pre>Arc</pre> and <pre>Value</pre>. One
                or more of them can be <pre>children</pre> of a <pre>Knob</pre>.
            </Description>
            <Grid>
                <PointerExample />
                <ArcExample />
                <ScaleExample />
                <ValueExample />
            </Grid>
            <H2>
                <span>Combinations</span>
            </H2>
            <Description>
                You can combine all of the base components in a knob.
            </Description>
            <Grid>
                <ArcPointerExample />
                <ScalePointerExample />
                <ArcPointerValueExample />
                <ScaleScale />
            </Grid>
            <H2>
                <span>Custom Elements</span>
            </H2>
            <Description>
                The <pre>Pointer</pre> and <pre>Scale</pre> components can also
                render custom components.
            </Description>
            <Grid>
                <PointerCustom />
                <ScaleCustom />
            </Grid>
            <H2>
                <span>Events</span>
            </H2>
            <Grid>
                <EventMouse />
                <EventReadOnly />
                <EventNoWheel />
            </Grid>
            <H2>
                <span>Styles</span>
            </H2>
            <Description>
                The components and the knob itself can be styled by addingthe{' '}
                <pre>className</pre> prop. The <pre>Scale</pre> can have an
                additional <pre>activeClassname</pre> prop to style the active tick
                differently. As all components are SVG you can also use filters in
                your styles.
            </Description>
            <Grid>
                <ArcPointerStyled />
                <ScaleStyled />
                <ScaleStyled2 />
            </Grid>
        </div>
    )
}
