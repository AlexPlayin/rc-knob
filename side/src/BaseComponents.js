import React from 'react'

import { H1, H2, Grid, Description } from './styled'
import PointerExample from './examples/Pointer'
import ScaleExample from './examples/Scale'
import ArcExample from './examples/Arc'
import ArcPointerExample from './examples/Arc_Pointer'
import ScalePointerExample from './examples/Scale_Pointer'
import ValueExample from './examples/Value'
import ArcPointerValueExample from './examples/Arc_Pointer_Value'
import ScaleScale from './examples/Scale_Scale'
import PointerCustom from './examples/Pointer_custom'
import ScaleCustom from './examples/Scale_custom'

export default () => (
    <div>
        <H1>React Knob</H1>
        <H2>
            <span>Base Components</span>
        </H2>
        <Description>
            The base of all knobs is the <pre>Knob</pre> component which handles
            all of the user interaction and calculation of the knob value.{' '}
            <br /> The display part is splitted into the <pre>Pointer</pre>,{' '}
            <pre>Scale</pre>, <pre>Arc</pre> and <pre>Value</pre>. One or more{' '}
            of them can be <pre>children</pre> of a <pre>Knob</pre>.
        </Description>
        <Grid>
            <PointerExample />
            <ArcExample />
            <ScaleExample />
            <ValueExample />
        </Grid>
        <H2>
            <span>Combinations</span>
        </H2>
        <Description>
            You can combine all of the base components in a knob.
        </Description>
        <Grid>
            <ArcPointerExample />
            <ScalePointerExample />
            <ArcPointerValueExample />
            <ScaleScale />
        </Grid>
        <H2>
            <span>Custom Elements</span>
        </H2>
        <Description>
            The <pre>Knob</pre> component and the <pre>Pointer</pre> and{' '}
            <pre>Scale</pre> components can also render custom components. All
            direct children of the <pre>Knob</pre> get the following props
            passed: <pre>percentage</pre> <pre>size</pre>, <pre>value</pre>,
            <pre>angleOffset</pre>,<pre>angleRange</pre>,{' '}
            <pre>radius: size / 2</pre>,<pre>center: size / 2</pre>,{' '}
            <pre>steps</pre>. The children of <pre>Pointer</pre> get{' '}
            <pre>width</pre>, <pre>height</pre>, <pre>percentage</pre> passed.
        </Description>
        <Grid>
            <PointerCustom />
            <ScaleCustom />
        </Grid>
    </div>
)
