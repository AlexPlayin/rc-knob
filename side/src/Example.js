import React from 'react'
import { Prism as SyntaxHighlighter } from 'react-syntax-highlighter'
import { H3, Cell, CodeBlock } from './styled'
import { Knob } from 'rc-knob'
import theme from './prismTheme'

const renderKnob = (child, options = {}) => (
    <Knob
        onChange={value => console.log(value)}
        size={100}
        value={0}
        angleOffset={220}
        angleRange={280}
        min={0}
        max={100}
        {...options}
        snap
    >
        {child}
    </Knob>
)

export default ({ example, knob, title, options }) => (
    <Cell>
        <H3>{title}</H3>
        {renderKnob(knob, options)}
        <CodeBlock>
            <SyntaxHighlighter language="jsx" style={theme}>
                {example}
            </SyntaxHighlighter>
        </CodeBlock>
    </Cell>
)
