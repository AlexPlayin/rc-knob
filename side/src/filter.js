import React from 'react'

export default () => (
    <svg>
        <filter id="inner-shadow">
            <feOffset dx="0" dy="5" />
            <feGaussianBlur stdDeviation="5" result="offset-blur" />
            <feComposite
                operator="out"
                in="SourceGraphic"
                in2="offset-blur"
                result="inverse"
            />
        </filter>
    </svg>
)
