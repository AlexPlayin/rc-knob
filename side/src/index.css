@import url('https://fonts.googleapis.com/css?family=Work+Sans:100,300');
@import url('https://fonts.googleapis.com/css?family=Eczar:800');

html {
    height: 100%;
}

body {
    margin: 0;
    padding: 0;
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto',
        'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans',
        'Helvetica Neue', sans-serif;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
    height: 100%;
    padding-top: 254px;
}

code {
    font-family: source-code-pro, Menlo, Monaco, Consolas, 'Courier New',
        monospace;
}
.value {
    font-family: 'Eczar', serif;
    fill: #fc5a96;
    font-size: 50px;
}
.value2 {
    font-family: 'Work Sans', sans-serif;
    fill: #fc5a96;
    font-size: 30px;
}
