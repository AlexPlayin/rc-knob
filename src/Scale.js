import React from 'react'

const renderCircle = ({
    tickWidth,
    translateX,
    translateY,
    angleOffset,
    stepSize,
    center,
    color,
    active,
    activeColor,
    activeClassName,
    className,
}) => (_, i) =>
    console.log('center', center) || (
        <circle
            r={tickWidth}
            key={i}
            className={i === active ? activeClassName : className}
            fill={i === active ? activeColor : color}
            stroke="none"
            transform={`
        rotate(${angleOffset + stepSize * i} ${center} ${center}) 
        translate(${translateX} ${translateY})
        `}
        />
    )

const renderRect = ({
    tickWidth,
    tickHeight,
    translateX,
    translateY,
    angleOffset,
    stepSize,
    center,
    color,
    active,
    activeColor,
    activeClassName,
    className,
}) => (_, i) => (
    <rect
        className={i === active ? activeClassName : className}
        fill={i === active ? activeColor : color}
        stroke="none"
        width={tickWidth}
        height={tickHeight}
        key={i}
        transform={`
        rotate(${angleOffset + stepSize * i} ${center} ${center}) 
        translate(${translateX} ${translateY})
        `}
    />
)
const renderCustom = ({ fn, ...props }) => (_, i) => fn({ ...props, i })

export const Scale = ({
    angleRange,
    steps,
    type = 'rect',
    radius,
    tickWidth,
    tickHeight,
    angleOffset,
    center,
    color,
    activeColor = color,
    className,
    activeClassName = className,
    fn,
    percentage,
}) => {
    const stepSize = angleRange / steps
    const length = steps + (angleRange === 360 ? 0 : 1)
    const translateX = center - tickWidth / 2
    const translateY = center - radius

    const active = Math.round((length - 1) * percentage)
    const renderFn =
        type === 'circle'
            ? renderCircle({
                  tickWidth,
                  translateX,
                  translateY,
                  center,
                  angleOffset,
                  stepSize,
                  color,
                  active,
                  activeColor,
                  className,
                  activeClassName,
              })
            : type === 'rect' && !fn
            ? renderRect({
                  tickWidth,
                  tickHeight,
                  translateX,
                  translateY,
                  angleOffset,
                  stepSize,
                  center,
                  color,
                  active,
                  activeColor,
                  className,
                  activeClassName,
              })
            : renderCustom({
                  fn,
                  tickWidth,
                  tickHeight,
                  translateX,
                  translateY,
                  angleOffset,
                  stepSize,
                  center,
                  color,
                  active,
                  activeColor,
                  className,
                  activeClassName,
              })
    return <g>{Array.from({ length }, renderFn)}</g>
}
