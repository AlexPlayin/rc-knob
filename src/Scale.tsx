import React from 'react';
import type { PropsWithKnobState } from 'types';

interface RenderProps {
    translateX: number;
    translateY: number;
    angleOffset: number;
    stepSize: number;
    center: number;
    color: string;
    className: string;
    active: number;
    activeColor: string;
    activeClassName: string;
}

const renderCircle =
    ({
        tickWidth,
        translateX,
        translateY,
        angleOffset,
        stepSize,
        center,
        color,
        active,
        activeColor,
        activeClassName,
        className,
    }: RenderProps & { tickWidth: number }) =>
    (_: any, i: number) =>
        (
            <circle
                r={tickWidth}
                key={i}
                className={i === active ? activeClassName : className}
                fill={i === active ? activeColor : color}
                stroke="none"
                transform={`
        rotate(${angleOffset + stepSize * i} ${center} ${center}) 
        translate(${translateX} ${translateY})
        `}
            />
        );

const renderRect =
    ({
        tickWidth,
        tickHeight,
        translateX,
        translateY,
        angleOffset,
        stepSize,
        center,
        color,
        active,
        activeColor,
        activeClassName,
        className,
    }: RenderProps & { tickWidth: number; tickHeight: number }) =>
    (_: any, i: number) =>
        (
            <rect
                className={i === active ? activeClassName : className}
                fill={i === active ? activeColor : color}
                stroke="none"
                width={tickWidth}
                height={tickHeight}
                key={i}
                transform={`
        rotate(${angleOffset + stepSize * i} ${center} ${center}) 
        translate(${translateX} ${translateY})
        `}
            />
        );

export interface RenderCustomProps extends RenderProps {
    i: number;
    tickWidth: number;
    tickHeight: number;
    steps: number;
    percentage: number;
}

const renderCustom =
    ({
        fn,
        ...props
    }: {
        fn: (props: RenderCustomProps) => void;
        tickWidth: number;
        tickHeight: number;
        steps: number;
        percentage: number;
    } & RenderProps) =>
    (_: any, i: number) =>
        fn({ ...props, i });

interface Props {
    angleRange: number;
    steps: number;
    type?: string;
    radius: number;
    tickWidth: number;
    tickHeight: number;
    color: string;
    activeColor?: string;
    className: string;
    activeClassName?: string;
    fn?: (props: RenderCustomProps) => void;
}

export const Scale = ({
    angleRange,
    steps,
    type = 'rect',
    radius,
    tickWidth,
    tickHeight,
    angleOffset,
    center,
    color,
    activeColor = color,
    className,
    activeClassName = className,
    fn,
    percentage,
}: PropsWithKnobState<Props>) => {
    const stepSize = angleRange / steps;
    const length = steps + (angleRange === 360 ? 0 : 1);
    const translateX = center - tickWidth / 2;
    const translateY = center - radius;
    if (percentage === null) {
        return <></>;
    }

    const active = Math.round((length - 1) * percentage);

    function getRenderFn() {
        if (type === 'circle') {
            return renderCircle({
                tickWidth,
                translateX,
                translateY,
                center,
                angleOffset,
                stepSize,
                color,
                active,
                activeColor,
                className,
                activeClassName,
            });
        }
        if (type === 'rect' && !fn) {
            return renderRect({
                tickWidth,
                tickHeight,
                translateX,
                translateY,
                angleOffset,
                stepSize,
                center,
                color,
                active,
                activeColor,
                className,
                activeClassName,
            });
        }
        if (fn) {
            if (percentage === null) {
                return <></>;
            }
            return renderCustom({
                fn,
                tickWidth,
                tickHeight,
                translateX,
                translateY,
                angleOffset,
                stepSize,
                center,
                color,
                active,
                activeColor,
                className,
                activeClassName,
                steps,
                percentage,
            });
        }
    }

    const renderFn = getRenderFn();

    // @ts-expect-error
    return <g>{Array.from({ length }, renderFn)}</g>;
};
