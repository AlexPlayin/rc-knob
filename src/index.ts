export * from './Knob'
export * from './Arc'
export * from './Scale'
export * from './Pointer'
export * from './Value'
export * from './Range'
export * from './Spiral'
export * from './Label'
export * from './types'
