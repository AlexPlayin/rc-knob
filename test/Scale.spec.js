import React from 'react'
import { shallow } from 'enzyme'
import { Scale } from '../src/Scale'

describe('Scale', () => {
    it('renders correct no type is set', () => {
        const component = shallow(
            <Scale
                angleRange={200}
                steps={5}
                radius={40}
                tickWidth={2}
                tickHeight={10}
                angleOffset={90}
                center={20}
                percentage={50}
                color="lime"
                activeColor="red"
                className="someClassName"
                activeClassName="someActiveClassName"
            />
        )
        expect(component).toMatchSnapshot()
    })
    it('renders correct no type is set to rect', () => {
        const component = shallow(
            <Scale
                type="rect"
                angleRange={200}
                steps={5}
                radius={40}
                tickWidth={2}
                tickHeight={10}
                angleOffset={90}
                center={20}
                percentage={50}
                color="lime"
                activeColor="red"
                className="someClassName"
                activeClassName="someActiveClassName"
            />
        )
        expect(component).toMatchSnapshot()
    })
    it('renders correct no type is set to circle', () => {
        const component = shallow(
            <Scale
                type="circle"
                angleRange={200}
                steps={5}
                radius={40}
                tickWidth={2}
                tickHeight={10}
                angleOffset={90}
                center={20}
                percentage={50}
                color="lime"
                activeColor="red"
                className="someClassName"
                activeClassName="someActiveClassName"
            />
        )
        expect(component).toMatchSnapshot()
    })
    it('renders correct with custom render function for ticks', () => {
        const fn = ({ tickWidth, translateX, translateY, i }) => (
            <circle r={tickWidth} x={translateX} y={translateY} key={i} />
        )
        const component = shallow(
            <Scale
                angleRange={200}
                steps={5}
                radius={40}
                tickWidth={2}
                tickHeight={10}
                angleOffset={90}
                center={20}
                percentage={50}
                color="lime"
                activeColor="red"
                className="someClassName"
                activeClassName="someActiveClassName"
                fn={fn}
            />
        )
        expect(component).toMatchSnapshot()
    })
})
