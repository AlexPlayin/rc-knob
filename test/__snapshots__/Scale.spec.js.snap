// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scale renders correct no type is set 1`] = `
<g>
  <rect
    className="someClassName"
    fill="lime"
    height={10}
    key="0"
    stroke="none"
    transform="
        rotate(90 20 20) 
        translate(19 -20)
        "
    width={2}
  />
  <rect
    className="someClassName"
    fill="lime"
    height={10}
    key="1"
    stroke="none"
    transform="
        rotate(130 20 20) 
        translate(19 -20)
        "
    width={2}
  />
  <rect
    className="someClassName"
    fill="lime"
    height={10}
    key="2"
    stroke="none"
    transform="
        rotate(170 20 20) 
        translate(19 -20)
        "
    width={2}
  />
  <rect
    className="someClassName"
    fill="lime"
    height={10}
    key="3"
    stroke="none"
    transform="
        rotate(210 20 20) 
        translate(19 -20)
        "
    width={2}
  />
  <rect
    className="someClassName"
    fill="lime"
    height={10}
    key="4"
    stroke="none"
    transform="
        rotate(250 20 20) 
        translate(19 -20)
        "
    width={2}
  />
  <rect
    className="someClassName"
    fill="lime"
    height={10}
    key="5"
    stroke="none"
    transform="
        rotate(290 20 20) 
        translate(19 -20)
        "
    width={2}
  />
</g>
`;

exports[`Scale renders correct no type is set to circle 1`] = `
<g>
  <circle
    className="someClassName"
    fill="lime"
    key="0"
    r={2}
    stroke="none"
    transform="
        rotate(90 20 20) 
        translate(19 -20)
        "
  />
  <circle
    className="someClassName"
    fill="lime"
    key="1"
    r={2}
    stroke="none"
    transform="
        rotate(130 20 20) 
        translate(19 -20)
        "
  />
  <circle
    className="someClassName"
    fill="lime"
    key="2"
    r={2}
    stroke="none"
    transform="
        rotate(170 20 20) 
        translate(19 -20)
        "
  />
  <circle
    className="someClassName"
    fill="lime"
    key="3"
    r={2}
    stroke="none"
    transform="
        rotate(210 20 20) 
        translate(19 -20)
        "
  />
  <circle
    className="someClassName"
    fill="lime"
    key="4"
    r={2}
    stroke="none"
    transform="
        rotate(250 20 20) 
        translate(19 -20)
        "
  />
  <circle
    className="someClassName"
    fill="lime"
    key="5"
    r={2}
    stroke="none"
    transform="
        rotate(290 20 20) 
        translate(19 -20)
        "
  />
</g>
`;

exports[`Scale renders correct no type is set to rect 1`] = `
<g>
  <rect
    className="someClassName"
    fill="lime"
    height={10}
    key="0"
    stroke="none"
    transform="
        rotate(90 20 20) 
        translate(19 -20)
        "
    width={2}
  />
  <rect
    className="someClassName"
    fill="lime"
    height={10}
    key="1"
    stroke="none"
    transform="
        rotate(130 20 20) 
        translate(19 -20)
        "
    width={2}
  />
  <rect
    className="someClassName"
    fill="lime"
    height={10}
    key="2"
    stroke="none"
    transform="
        rotate(170 20 20) 
        translate(19 -20)
        "
    width={2}
  />
  <rect
    className="someClassName"
    fill="lime"
    height={10}
    key="3"
    stroke="none"
    transform="
        rotate(210 20 20) 
        translate(19 -20)
        "
    width={2}
  />
  <rect
    className="someClassName"
    fill="lime"
    height={10}
    key="4"
    stroke="none"
    transform="
        rotate(250 20 20) 
        translate(19 -20)
        "
    width={2}
  />
  <rect
    className="someClassName"
    fill="lime"
    height={10}
    key="5"
    stroke="none"
    transform="
        rotate(290 20 20) 
        translate(19 -20)
        "
    width={2}
  />
</g>
`;

exports[`Scale renders correct with custom render function for ticks 1`] = `
<g>
  <circle
    key="0"
    r={2}
    x={19}
    y={-20}
  />
  <circle
    key="1"
    r={2}
    x={19}
    y={-20}
  />
  <circle
    key="2"
    r={2}
    x={19}
    y={-20}
  />
  <circle
    key="3"
    r={2}
    x={19}
    y={-20}
  />
  <circle
    key="4"
    r={2}
    x={19}
    y={-20}
  />
  <circle
    key="5"
    r={2}
    x={19}
    y={-20}
  />
</g>
`;
